File: envelopes.clj
;; Download from https://clojure.org/community/downloads
;; alias clj="java -cp ~/Code/clojure-1.8.0/clojure-1.8.0.jar:. clojure.main"
;; $ clj -m envelopes envelopes.clj
;; or 
;; $ brew install leiningen
;; $ lein repl
;; user=> (load-file "envelopes.clj")
;; user=> (envelopes/-main [])

(ns envelopes)

(def NUM_TRIALS 10000)
(def PRIOR_LOWER_MAX 100)

(defn get-trial-values 
  "Helper to generate the [picked, other] list for a given run."
  [envelope lower_value]
    (let [values [lower_value (* 2 lower_value)]]
      (if (== envelope 0) values (reverse values))))

;; Returns the result of a single trial. We switch if the value is below the
;; cutoff.
(defn single-trial [cutoff]
  (let [values (get-trial-values (rand-int 2) (rand PRIOR_LOWER_MAX))]
    (if (>= (first values) cutoff) (first values) (second values))))

;; Returns the total value of all of the trials for a given cutoff level.       
(defn get-multi-trial-total [cutoff]
  (loop [trial 0 total 0]
    (if (>= trial NUM_TRIALS)
      total
      (recur (+ trial 1) (+ total (single-trial cutoff))))))
    
;; Computes the average value among many trials.  
(defn multi-trial [cutoff] (/ (get-multi-trial-total cutoff) NUM_TRIALS))

;; Prints the expected value for each possible cutoff.
(defn -main [& args]
  (loop [cutoff 0]
    (if (> cutoff (* 2 PRIOR_LOWER_MAX))
      ()
      (let [expected_value (multi-trial cutoff)]
        (printf "cutoff=%s, expected_value=%s\n" cutoff expected_value)
        (recur (+ cutoff 1))))))

File: envelopes.erl
% $ brew install erlang
% $ erl -compile envelopes
% $ erl -noshell -s envelopes main -s init stop
% or
% $ erl
% 1> c(envelopes).
% 2> envelopes:main().

-module(envelopes).
-export([main/0]).

-define(PRIOR_LOWER_MAX, 100).
-define(NUM_TRIALS, 10000).

% Helper to generate the [picked, other] list for a given run.
get_trial_values(1, LowerValue) -> [LowerValue, 2 * LowerValue];
get_trial_values(2, LowerValue) -> [2 * LowerValue, LowerValue].

% Returns the result of a single trial. We switch if the value is below the
% cutoff.
single_trial(Cutoff) ->
  [Choice, Other] = get_trial_values(rand:uniform(2), rand:uniform() * ?PRIOR_LOWER_MAX),
  if Choice >= Cutoff -> Choice; true -> Other end.

% Returns the total value of all of the trials for a given cutoff level.
get_multi_trial_total(_, 0, Total) -> Total;
get_multi_trial_total(Cutoff, Remaining, Total) ->
  get_multi_trial_total(Cutoff, Remaining - 1, Total + single_trial(Cutoff)).

% Computes the average value among many trials.
multi_trial(Cutoff) ->
  get_multi_trial_total(Cutoff, ?NUM_TRIALS, 0) / ?NUM_TRIALS.

% Prints the expected value for each possible cutoff.
cutoff_trials(_, 0) -> ok;
cutoff_trials(Cutoff, Remaining) ->
  ExpectedValue = multi_trial(Cutoff),
  io:format("cutoff=~w, expected_value=~w~n", [Cutoff, ExpectedValue]),
  cutoff_trials(Cutoff + 1, Remaining - 1).

main() ->
  cutoff_trials(0, 2 * ?PRIOR_LOWER_MAX + 1).
